### ISSUE TITLE
Interpolation errors in bcdec__smooth_alpha_block

### ISSUE THREAD

**Issue by jandk** [iOrange/bcdec#12]
In this function you interpolate between the two different gray values.

I was checking the code, and it seems that your rounding is off.

In the divide by 7 part, the addition is 1, should be 3.
In the divide by 5 part, the addition is 1, should be 2.

I compared these with full float unpacking and packing.

**Comment by iOrange** [comment]
Hi!
Thanks for checking out my small library.
I see that you've successfully adopted bc1-bc3 functions for your [valen](https://github.com/jandk/valen/tree/main/valen-app/src/main/java/be/twofold/valen/converter/decoder) application, nice!

I don't believe that that part is wrong tbh, it was written against specification and I've compared different decompression results to other "reference" decompressors.

Here's the list of official specifications I was using to write my code:
https://learn.microsoft.com/en-us/windows/win32/direct3d10/d3d10-graphics-programming-guide-resources-block-compression#bc3
https://www.khronos.org/opengl/wiki/S3_Texture_Compression#DXT5_Format
https://en.wikipedia.org/wiki/S3_Texture_Compression#DXT4_and_DXT5

I've also made a simple test to compare both my integer interpolation and more precise floating point one, and rounding error is negligible (esp. given the speed difference)
https://onlinegdb.com/PMqHzmzeS

If you believe it is still wrong, please provide and example and I'm gladfully check it and fix my code of necessary.

Thanks,
Sergii

**Comment by jandk** [comment]
Hey @iOrange, thanks for responding.

I might have to double check, but I basically compared full float operations (so toFloat, lerp, toUnorm) to what you did and it does seem for me it does some strange rounding. I'll post the code here later (on different PC now).

Now what I did see is that a lot of implementations basically declare what needs to be done on the GPU, and what rounding errors to expect (and it seems dx is pretty liberal with these.

I was reading this spec here
https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#19.5%20Block%20Compression%20Formats

Linked to me by fgiesen (https://fgiesen.wordpress.com/2021/10/04/gpu-bcn-decoding/), or at least he talks about the different specs. Note that this is an old article, but I looked and I couldn't really find definitive information on S3TC from Khronos.

Stuff I found out so far: for BC1 565 it's ok to bit replicate (as in 19.5.3) but for BC3/BC4U/BC5U they specify the interpolation in floating point format. That's what I did for now, and that's the version that I compared. 

Let me whip up an example, and I'll show you. Tbh at this point it seems a bit pedantic, but I went down a rabbithole nobody seems to know the answer for because all the specs kinda seem to disagree :D

**Comment by iOrange** [comment]
Well yeah I agree, the "pre-BC7" times are a bit weird - the spec didn't have any hard requirements for a bit-precise decoding, that 's why different implementations (even hardware ones) do it a bit differently.

One shortfall of my code that I'm aware of - is not handling the signed mode for the BC4/BC5, I'll fix that soon.
But as for the alpha interpolation - not sure I'm convinced yet.

Will gladly check your example as I'm all for improving the code for the benefit of all users.

**Comment by jandk** [comment]
Cleaned up my proof of concept.

 http://tpcg.io/_APMKQS 

And yes, I also still have to do signed mode :D

**Comment by iOrange** [comment]
Okay, I guess I see your point now.
Our difference is rounding convention.
I was having a "C-lang float to uint8" in mind, where it just drops any fractional part.
You're pointing to the  section 3.2.3.6  "FLOAT -> UNORM", where it states:
```
Convert to integer:
   c = c + 0.5f.
   Drop the decimal fraction, and the remaining floating point (integral) value is converted directly to an integer.
```
In this case yes, you're right.   I have to test this better, but now you've convinced me to dig deeper, thanks!

**Comment by jandk** [comment]
You're welcome! I'm just trying to understand to the best of my knowledge, but it's all very confusing.

I saw you did the * 527 + 23 >> 6 conversions for BC1 as well, which are actually floating point accurate. But I'm wondering if you (not you, but as in someone who implements BC decoders) should go for fp accuracy in general, or be happy that you're off by a bit. I don't know :)

Also those constants are magic, did you find an explanation anywhere on the internet?

Sorry for the rambling :)

**Comment by iOrange** [comment]
Yeah, so the 565 -> 888 conversion is just my carry-over code from my other (older) projects. I don't remember exactly where did I find that magic, I just know that it works better than simple bits extension (btw this is what that 19.5.3 section suggests) while still being super fast.

As for the bc3_alpha/bc4 code - I wrote it against the spec, compared the output to the "reference" decoder and against my own floating-point implementation, tinkered a bit and put a "good enough" on it.

But I guess I can do better precision-wise, just need to step carefully - I personally don't like performance regressions in libraries, and would like to not cause one in my own, especially if the output quality benefits are miniscule.

But I will investigate potential precision improvements  (as well as a signed bc4/bc5 mode heh).

**Comment by richgel999** [comment]
On rounding during interpolation for BC3 alpha/BC4/BC5 - Intel GPU's don't round AFAIK. My code (like in crunch) doesn't round. See pages 123-124 of "Intel® 965 Express Chipset, Family and Intel® G35 Express, Chipset Graphics Controller PRM" for example. Not sure what other GPU's do.

https://www.x.org/docs/intel/VOL_1_graphics_core.pdf

![image](https://github.com/iOrange/bcdec/assets/6934809/e6ab5bc8-30ce-4f7c-b857-c2cc1b20058f)

**Comment by iOrange** [comment]
@richgel999 didn't expect an expert of your level to appear on my humble home project, truly an honour for me.
And yes, I'm still on the fence on this issue. Like I'm thinking of adding a compile-time option (a define) for a different behaviour for those who really need it, but leaving the default codepath untouched.

**Comment by jandk** [comment]
I have a feeling at this point it's more of an intellectual exercise than anything else. I like your idea @iOrange, turning it into a compile time define. There's no definitive spec anywhere on this issue (except for BC6H and BC7), and all methods I've tried seem to differ only by one value at most.

For my project I think I might decide to just keep it fully in float, maybe render it out to a couple of tables, no idea yet. I don't need raw speed anyway.

**Comment by jandk** [comment]
FYI, I did finally manage to release my own library. Through some quite extensive benchmarking I found out that the float version (in Java) is actually faster than the int multiply/shift with the magic constants etc. 

Decided to lerp with some fma, and it seems to be quite fast.

I've also decided to implement BC6H using a "microcode" system, where the operations all contain the number of bits, the shift, the color to apply it to and a bit to reverse the read bits if needed. I really like this approach, it saves me from having lists of function calls. Might be an idea for you?

For me you may close this bug, as it's obviously fixed by the above PRs.

I'll leave a link if you want to check my implementation out.

https://github.com/jandk/tinybcdec

**Comment by iOrange** [comment]
Finally I had time to get back to this, and after thinking on this, decided next:

the BC3 alpha decoding will do truncation - most of the older "specs" point to this behaviour
added a compile-time option to enable separate BC4/BC5 decoding functions that do proper rounding (both integer and float) + added signed/unsigned mode.

963c5e56b7a335e066cff7d16a3de75f4e8ad366

**Comment by jandk** [comment]
Sorry to bring this topic back up, but I saw your commit from a few days ago, changing the bc4 interpolation by removing the addition constants in the expressions, and having a float accurate version on top of it.

I might've told this before, or it might not have been clear, or it might have gotten lost, but it's very easy to adjust the factors to have actual float accurate conversions.

I've done `((7 - x)*a0 + x*a1 + 3) / 7` and `((5-x)*a0 + x*a1 + 2) / 5` and they match exactly to doing a float round trip.

I thought maybe I was doing something wrong, so I wrote a test case you can look at here:
https://github.com/jandk/tinybcdec/blob/main/src/test/java/be/twofold/tinybcdec/BC4UInterpolationTest.java

This should simplify your code a bit, and could get rid of the compile time flag.

Edit: I see what you did. I like that the approach works for both signed and unsigned. My for now, only works for unsigned, for signed I go full float, might have to revisit this in the future.

**Comment by iOrange** [comment]
Hi @jandk , I appreciate all your input, truly.
I'm not 100% sure what's the best way, so I decided to leave BC3 alpha decoding to be "old spec" and make a dedicated BC4/BC5 precise decoding functions as they are more modern formats.

In any case - this was initially my sandbox for understanding BC compression (mostly BC6H/BC7) that I decided to share with the world, and it's completely Public Domain so if anyone feels the need to do proper rounding for BC3 alpha - it's as easy as to replace `bcdec__smooth_alpha_block` with `bcdec__bc4_block`.

I'm thinking of exploring the land of ASTC and PVRTC next which looks to be a lot of fun :)
